testsuite/certificates/__init__.py

```diff
@@ -0,0 +1,158 @@
+                         *args: str,
+                         stdin: Optional[str] = None,
+                         env: Optional[Dict[str, str]] = None):
+        args = (self.binary, command, *args)
+        try:
+            response = subprocess.run(args,
+                                      stderr=subprocess.PIPE,
+                                      stdout=subprocess.PIPE,
+                                      input=stdin,
+                                      universal_newlines=bool(stdin),
+                                      check=False,
+                                      env=env)
+            if response.returncode != 0:
+                raise CFSSLException(f"CFSSL command {args} returned non-zero response code, error {response.stderr}")
+            return json.loads(response.stdout)
+        except Exception as exception:
+            # If some error occurs, first check if the binary exists to throw better error
+            if not self.exists:
+                raise AttributeError("CFSSL binary does not exist") from exception
+            raise exception
+
+    @cached_property
+    def exists(self):
+        """Returns true if the binary exists and is correctly set up"""
+        return shutil.which(self.binary)
+
+    def generate_key(self, common_name: str, names: Optional[List[Dict[str, str]]] = None,
+                     hosts: Optional[List[str]] = None) -> UnsignedKey:
+        """Generates unsigned key"""
+        data: Dict[str, Any] = {
+            "CN": common_name
+        }
+        if names:
+            data["names"] = names
+        if hosts:
+            data["hosts"] = hosts
+
+        result = self._execute_command("genkey", "-", stdin=json.dumps(data))
+        return UnsignedKey(key=result["key"], csr=result["csr"])
+
+    def sign_intermediate_ca(self, key: UnsignedKey, certificate_authority: Certificate) -> Certificate:
+        """Signs intermediate ca"""
+        args = [
+            "-ca=env:CA",
+            "-ca-key=env:KEY",
+            f"-config={resources.files('testsuite.resources.tls').joinpath('intermediate_config.json')}"
+        ]
+        result = self._execute_command("sign", *args, "-", stdin=key.csr, env={
+            "CA": certificate_authority.certificate,
+            "KEY": certificate_authority.key})
+        return Certificate(key=key.key, certificate=result["cert"])
+
+    def sign(self, key: UnsignedKey, certificate_authority: Certificate) -> Certificate:
+        """Signs unsigned key"""
+        result = self._execute_command("sign", "-ca=env:CA", "-ca-key=env:KEY", "-", stdin=key.csr, env={
+            "CA": certificate_authority.certificate,
+            "KEY": certificate_authority.key})
+        return Certificate(key=key.key, certificate=result["cert"])
+
+    def generate_ca(self, common_name: str,
+                    names: List[Dict[str, str]],
+                    hosts: List[str]) -> Tuple[Certificate, UnsignedKey]:
+        """Generates CA"""
```

Comment:
Based on official documentation, I would extend the description to something like this: "Generates self-signed root CA certificate and private key"

Reply:
Updated